xdem/coreg/affine.py

```diff
@@ -54,147 +54,728 @@
+                "Argument `bin_sizes` must be an integer, or a dictionary of integers or iterables, "
+                "got {}.".format(type(bin_sizes))
+            )
+
+        if not callable(bin_statistic):
+            raise TypeError(
+                "Argument `bin_statistic` must be a function (callable), " "got {}.".format(type(bin_statistic))
+            )
+
+
+def _iterate_method(
+    method: Callable[..., Any],
+    iterating_input: Any,
+    constant_inputs: tuple[Any, ...],
+    tolerance: float,
+    max_iterations: int,
+    verbose: bool = False,
+) -> Any:
+    """
+    Function to iterate a method (e.g. ICP, Nuth and Kääb) until it reaches a tolerance or maximum number of iterations.
+
+    :param method: Method that needs to be iterated to derive a transformation. Take argument "inputs" as its input,
+        and outputs three terms: a "statistic" to compare to tolerance, "updated inputs" with this transformation, and
+        the parameters of the transformation.
+    :param iterating_input: Iterating input to method, should be first argument.
+    :param constant_inputs: Constant inputs to method, should be all positional arguments after first.
+    :param tolerance: Tolerance to reach for the method statistic (i.e. maximum value for the statistic).
+    :param max_iterations: Maximum number of iterations for the method.
+    :param verbose: Whether to print progress.
+
+    :return: Final output of iterated method.
     """
-    # Old implementation
-    # # Calculate the gradient of the slope
-    gradient_y, gradient_x = np.gradient(dem)
-    slope_tan = np.sqrt(gradient_x**2 + gradient_y**2)
-    aspect = np.arctan2(-gradient_x, gradient_y)
-    aspect += np.pi
-
-    # xdem implementation
-    # slope, aspect = xdem.terrain.get_terrain_attribute(
-    #     dem, attribute=["slope", "aspect"], resolution=1, degrees=False
-    # )
-    # slope_tan = np.tan(slope)
-    # aspect = (aspect + np.pi) % (2 * np.pi)
 
-    return slope_tan, aspect
+    # Initiate inputs
+    new_inputs = iterating_input
 
+    # Iteratively run the analysis until the maximum iterations or until the error gets low enough
+    # If verbose is True, will use progressbar and print additional statements
+    pbar = trange(max_iterations, disable=not verbose, desc="   Progress")
+    for i in pbar:
 
-def get_horizontal_shift(
-    elevation_difference: NDArrayf, slope: NDArrayf, aspect: NDArrayf, min_count: int = 20
-) -> tuple[float, float, float]:
+        # Apply method and get new statistic to compare to tolerance, new inputs for next iterations, and
+        # outputs in case this is the final one
+        new_inputs, new_statistic = method(new_inputs, *constant_inputs)
+
+        # Print final results
+        # TODO: Allow to pass a string to _iterate_method on how to print/describe exactly the iterating input
+        if verbose:
+            pbar.write(f"      Iteration #{i + 1:d} - Offset: {new_inputs}; Magnitude: {new_statistic}")
```

Comment:
I would need to test it to have a better idea of what the prints look like, e.g. for NuthKaab, but indeed generic statements can make it less easy to understand for the user.

Reply:
Yes, opening an issue to get that done.

Reply:
Opened #579